File: Dev/C++/Mascaret/src/VEHA/Behavior/Common/BehaviorExecution.cpp
#include "VEHA/Behavior/Common/BehaviorExecution.h"
#include "VEHA/Behavior/Common/Behavior.h"
#include "VEHA/Behavior/Time/TimeExpression.h"

namespace VEHA
{

BehaviorExecution::BehaviorExecution(shared_ptr<Behavior> specif,shared_ptr<InstanceSpecification> host,const Parameters& p, bool sync)
:_specification(specif),_host(host),_sync(sync),_isFinished(false),_isSuspended(false),parameters(p),result()
{
}

BehaviorExecution::~BehaviorExecution()
{
}

void BehaviorExecution::stop(void)
{
	if(!_isFinished)
	{
		_isFinished=true;
		_cbOnBehaviorStop(this);
	}
}
void BehaviorExecution::pause(void)
{
	if(!_isSuspended)
	{
		_isSuspended=true;
		_cbOnBehaviorSuspended(this);
	}
}
void BehaviorExecution::restart(void)
{
	if(_isSuspended)
	{
		_isSuspended=false;
		_cbOnBehaviorRestarted(this);
	}
}
void BehaviorExecution::addCallbackOnBehaviorStart(const boost::function1<void, BehaviorExecution*> & f)
{
	_cbOnBehaviorStart.connect(f);
}

void BehaviorExecution::removeCallbackOnBehaviorStart(const boost::function1<void, BehaviorExecution*>& f)
{
	_cbOnBehaviorStart.disconnect(&f);
}

void BehaviorExecution::addCallbackOnBehaviorStop(const boost::function1<void, BehaviorExecution*> & f)
{
	_cbOnBehaviorStop.connect(f);
}

void BehaviorExecution::removeCallbackOnBehaviorStop(const boost::function1<void, BehaviorExecution*>& f)
{
	_cbOnBehaviorStop.disconnect(&f);
}
void BehaviorExecution::addCallbackOnBehaviorSuspended(const boost::function1<void, BehaviorExecution*> & f)
{
	_cbOnBehaviorSuspended.connect(f);
}

void BehaviorExecution::removeCallbackOnBehaviorSuspended(const boost::function1<void, BehaviorExecution*>& f)
{
	_cbOnBehaviorSuspended.disconnect(&f);
}
void BehaviorExecution::addCallbackOnBehaviorRestarted(const boost::function1<void, BehaviorExecution*> & f)
{
	_cbOnBehaviorRestarted.connect(f);
}
void BehaviorExecution::removeCallbackOnBehaviorRestarted(const boost::function1<void, BehaviorExecution*>& f)
{
	_cbOnBehaviorRestarted.disconnect(&f);
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Behavior/Common/BehaviorExecution.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public abstract class BehaviorExecution : System.IEquatable<BehaviorExecution>
    {
        private Behavior specification;
        public Behavior Specification
        {
            get { return specification; }
            set { specification = value; }
        }

        private InstanceSpecification host;
        public InstanceSpecification Host
        {
            get { return host; }
            set { host = value; }
        }

        private bool sync;
        public bool Sync
        {
            get { return sync; }
            set { sync = value; }
        }

        private TimeExpression start;
        public TimeExpression Start
        {
            get { return start; }
            set { start = value; }
        }


        private TimeExpression finish;
        public TimeExpression Finish
        {
            get { return finish; }
            set { finish = value; }
        }


        private bool isFinished;
        public bool IsFinished
        {
            get { return isFinished; }
            set { isFinished = value; }
        }


        private bool isSuspended;
        public bool IsSuspended
        {
            get { return isSuspended; }
            set { isSuspended = value; }
        }


        private TimeExpression lastCalledTime;
        public TimeExpression LastCalledTime
        {
            get { return lastCalledTime; }
            set { lastCalledTime = value; }
        }

        public Dictionary<String, ValueSpecification> parameters;


        public List<ValueSpecification> result;

        public delegate void OnBehaviorStopDelegate(BehaviorExecution be);
        private List<OnBehaviorStopDelegate> cbStop = new List<OnBehaviorStopDelegate>();
        public void addCallbackOnBehaviorStop(OnBehaviorStopDelegate fct)
        {
            cbStop.Add(fct);
        }

        protected void onBehaviorStop()
        {
            foreach (OnBehaviorStopDelegate fct in cbStop)
                fct(this);
        }

        public BehaviorExecution(Behavior behavior, InstanceSpecification host, Dictionary<string, ValueSpecification> p, bool sync)
        {
            specification = behavior;
            this.host = host;
            this.sync = sync;
            isFinished = false;
            isSuspended = false;
            parameters = p;
            result = new List<ValueSpecification>();
        }

        public BehaviorExecution()
        {

        }

        public virtual void init(Behavior behavior, InstanceSpecification host, Dictionary<string, ValueSpecification> p, bool sync)
        {
            specification = behavior;
            this.host = host;
            this.sync = sync;
            isFinished = false;
            isSuspended = false;
            parameters = p;
            result = new List<ValueSpecification>();
        }

        public bool Equals(BehaviorExecution other)
        {
            return this == other;
        }

        public virtual void stop()
        {

            if (!isFinished)
            {
                isFinished = true;
                onBehaviorStop();
            }

        }
        public virtual void pause()
        {
            if (!isSuspended)
            {
                isSuspended = true;
                //cbOnBehaviorRestarted
            }

        }

        public virtual void restart()
        {
            if (isSuspended)
            {
                isSuspended = false;
                //cbOnBehaviorRestarted
            }
        }

        public abstract double execute(double dt);

        //callbacks functions
    }
}

